### ISSUE TITLE
Is the static still not working?

### ISSUE THREAD

**Issue by s2s-k** [ValveSoftware/GameNetworkingSockets#225]
As far as I understood, in 1.4.0 you fixed the installation of the static version via vcpkg... or not?

```powershell
PS E:\Work\vcpkg> ./vcpkg search gamenetworkingsockets
gamenetworkingsockets    1.4.0            GameNetworkingSockets is a basic transport layer for games.
The result may be outdated. Run `git pull` to get the latest results.

If your port is not listed, please open an issue at and/or consider making a pull request:
    https://github.com/Microsoft/vcpkg/issues
PS E:\Work\vcpkg> ./vcpkg install gamenetworkingsockets:x64-windows-static
Computing installation plan...
Error: gamenetworkingsockets[core] is only supported on '!(static & windows) & !uwp & !(arm64 & windows)'
```

I saw this PR [[gamenetworkingsockets] Enable building as a static library, and add webrtc feature](https://github.com/microsoft/vcpkg/pull/21759/files#diff-df79355ed795c21b49d5d8596e433a23b90e4b326d746a7beaaccf19eb9ebe58R8), it was not included in 1.4.0? Your [PR 1.4.0](https://github.com/zpostfacto/vcpkg/blob/918d66274069b812343cd4cbf36abd9150522ade/ports/gamenetworkingsockets/vcpkg.json#L7) still does not support the static version. Why?

**Comment by eddiejames** [comment]
You can try with this vcpkg patch (also includes some physx stuff i'm working on): https://github.com/eddiejames/vcpkg/commit/420290091e82288e461473e3447e6e4901f0c8df

My original PR wasn't accepted because of webrtc stuff... I'll try and resubmit my change sometime

**Comment by s2s-k** [comment]
> You can try with this vcpkg patch

I applied this patch, GNS compiled statically, but my project compiles with errors of unresolved external symbols.

```
error LNK2019: unresolved external symbol __imp_GameNetworkingSockets_Init referenced in function main
```

I have enabled the use of static vcpkg libraries in the project properties. I don't know why vcpkg didn't include the library in my project. Can this be done manually?

**Comment by s2s-k** [comment]
Okay, I solved this problem with a macro.

```cpp
#define STEAMNETWORKINGSOCKETS_STATIC_LINK
```

But now there are the same errors in libcrypto.lib.

```
error LNK2019: unresolved external symbol __imp_DeregisterEventSource referenced in function OPENSSL_showfatal
error LNK2019: unresolved external symbol __imp_RegisterEventSourceW referenced in function OPENSSL_showfatal
error LNK2019: unresolved external symbol __imp_ReportEventW referenced in function OPENSSL_showfatal
...
```

If I can solve this problem too, I'll write.

**Comment by s2s-k** [comment]
I did it. Thank you for your patch. The problem I wrote about above is solved by using additional Windows API libraries.

```cpp
#pragma comment(lib, "Advapi32.lib")
#pragma comment(lib, "User32.lib")
#pragma comment(lib, "Crypt32.lib")
```

It is also required to use the /MTd flag for Debug configuration and /MD for Release configuration.

**Comment by s2s-k** [comment]
Also a little warning, for those who will be switching to the static version. If you call `GameNetworkingSocket_Init` in the global object constructor, it may crash during static linking when it worked fine in the dynamic because of the different order of object initialization.

**Comment by bugproof** [comment]
Static should be a preferred way to consume GameNetworkingSockets. Dll makes it easier for reverse engineers to hack your game.

**Comment by rjayne-git** [comment]
I am also blocked on proper support for VCPKG_LIBRARY_LINKAGE static. I hope it is coming soon.

**Comment by zpostfacto** [comment]
This should be fixed now.  I'll close this bug when the new release is added to vcpkg.

**Comment by zpostfacto** [comment]
I'm hoping that this is resolved with v1.6.0.  If no, please tell me the exact way that you are building and the error messages you are seeing.
See also: https://github.com/microsoft/vcpkg/pull/52187

**Comment by Rewriter00x** [comment]
@zpostfacto 
I've been trying to build statically with /MT on windows. I've tried the following:
1. Building directly from my project's cmake after installing all vcpkg dependencies. It resulted into fild in finding dependencies on GNS
```
cmake_minimum_required(VERSION 3.16)
project(GNS_Static)

set(CMAKE_CXX_STANDARD 20)

if(WIN32 AND NOT MSVC)
    message(FATAL_ERROR "Only MSVC is supported on Windows.")
endif()

if (MSVC)
    set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug>)
    add_compile_options(/arch:AVX2)
endif()

set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)

add_subdirectory(dep/GameNetworkingSockets)

add_executable(${CMAKE_PROJECT_NAME} src/example_chat.cpp)

target_include_directories(${CMAKE_PROJECT_NAME} PUBLIC
        ${CMAKE_CURRENT_SOURCE_DIR}/src
)

target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE
        GameNetworkingSockets::static
)

target_compile_definitions(${CMAKE_PROJECT_NAME} PUBLIC
        STEAMNETWORKINGSOCKETS_STATIC_LINK
)
```
Error I got:
```
Could NOT find Protobuf (missing: Protobuf_LIBRARIES Protobuf_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files/JetBrains/CLion 2026.1.2/bin/cmake/win/x64/share/cmake-4.2/Modules/FindPackageHandleStandardArgs.cmake:654 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/JetBrains/CLion 2026.1.2/bin/cmake/win/x64/share/cmake-4.2/Modules/FindProtobuf.cmake:1135 (find_package_handle_standard_args)
  dep/GameNetworkingSockets/src/CMakeLists.txt:14 (find_package)
```

2. Building GameNetworkingSockets_s.lib separatelly, according to build.md. 
```
cmake_minimum_required(VERSION 3.16)
project(GNS_Static)

set(CMAKE_CXX_STANDARD 20)

if(WIN32 AND NOT MSVC)
    message(FATAL_ERROR "Only MSVC is supported on Windows.")
endif()

if (MSVC)
    set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug>)
    add_compile_options(/arch:AVX2)
endif()

set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)

add_executable(${CMAKE_PROJECT_NAME} src/example_chat.cpp)

target_include_directories(${CMAKE_PROJECT_NAME} PUBLIC
        ${CMAKE_CURRENT_SOURCE_DIR}/src
        dep/GameNetworkingSockets/include
)

target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE
        ${CMAKE_CURRENT_SOURCE_DIR}/dep/GameNetworkingSockets/lib/$<$<CONFIG:Debug>:Debug>$<$<CONFIG:Release>:Release>/GameNetworkingSockets_s.lib
)

target_compile_definitions(${CMAKE_PROJECT_NAME} PUBLIC
        STEAMNETWORKINGSOCKETS_STATIC_LINK
)
```
In that case linking dll version of the built library works, but GNS and it's dependencies require /MD linking with CRT. Linking static lib to my project and running example_chat.cpp produces 186 unresolved externals

3. Same as option 2, but building GNS with x64-windows-static triple, Protobuf_USE_STATIC_LIBS set to ON and using /MT for the GNS itself. In that case building shared lib of GNS results into errors:
```
Creating library src\GameNetworkingSockets.lib and object src\GameNetworkingSockets.exp
libprotobufd.lib(descriptor.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
libprotobufd.lib(any_lite.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
libprotobufd.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(match.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
libprotobufd.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
absl_log_internal_message.lib(log_message.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_search_1
bin\GameNetworkingSockets.dll : fatal error LNK1120: 4 unresolved externals
ninja: build stopped: subcommand failed.
```
Static lib build successfully, but results in 120 unresolved externals. They come from both protobuf and openssl
Also tried the same config with setting a bunch of vcpkg variables, but the result was the same
```
set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug> CACHE STRING "" FORCE)
set(VCPKG_TARGET_TRIPLET "x64-windows-static" CACHE STRING "" FORCE)

set(VCPKG_TARGET_ARCHITECTURE x64 CACHE STRING "" FORCE)
set(VCPKG_CRT_LINKAGE static CACHE STRING "" FORCE)
set(VCPKG_LIBRARY_LINKAGE static CACHE STRING "" FORCE)
set(VCPKG_CMAKE_SYSTEM_NAME WindowsStore CACHE STRING "" FORCE)

```

I have also tried building manually, but I still always get the linking errors from protobuf and openssl

In conclusion, the only working option I have found for now is to build GNS as a dll, have all dependents as dlls and link CRT as /MD, while I was trying to only have static libs and /MT CRT linkage. Will appreciate the help!

**Comment by Rewriter00x** [comment]
Update: I have tried to also link everything from GameNetworkingSockets\build\vcpkg_installed\x64-windows-static\lib to my cmake
```
cmake_minimum_required(VERSION 3.16)
project(GNS_Static)

set(CMAKE_CXX_STANDARD 20)

if(WIN32 AND NOT MSVC)
    message(FATAL_ERROR "Only MSVC is supported on Windows.")
endif()

if (MSVC)
    set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug>)
    add_compile_options(/arch:AVX2)
endif()

set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)

file(GLOB_RECURSE MY_SRC CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/src/*.cpp)

add_executable(${CMAKE_PROJECT_NAME} ${MY_SRC})

target_include_directories(${CMAKE_PROJECT_NAME} PUBLIC
        ${CMAKE_CURRENT_SOURCE_DIR}/src
        dep/GameNetworkingSockets/include
)

file(GLOB_RECURSE PB_LIB CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/dep/GameNetworkingSockets/build/vcpkg_installed/x64-windows-static/lib/*.lib)

target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE
        ${CMAKE_CURRENT_SOURCE_DIR}/dep/GameNetworkingSockets/lib/$<$<CONFIG:Debug>:Debug>$<$<CONFIG:Release>:Release>/GameNetworkingSockets_s.lib
        ${PB_LIB}
)

target_compile_definitions(${CMAKE_PROJECT_NAME} PUBLIC
        STEAMNETWORKINGSOCKETS_STATIC_LINK
)
```

That helped cleanup almost all unresolved externals, leaving only the next ones:
```
absl_log_internal_message.lib(log_message.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
libprotobuf-lite.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertOpenStore
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertCloseStore
libcrypto.lib(libdefault-lib-winstore_store.obj) : error LNK2001: unresolved external symbol __imp_CertCloseStore
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertEnumCertificatesInStore
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertFindCertificateInStore
libcrypto.lib(libdefault-lib-winstore_store.obj) : error LNK2001: unresolved external symbol __imp_CertFindCertificateInStore
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertDuplicateCertificateContext
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertFreeCertificateContext
libcrypto.lib(libdefault-lib-winstore_store.obj) : error LNK2001: unresolved external symbol __imp_CertFreeCertificateContext
libcrypto.lib(libcrypto-lib-e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertGetCertificateContextProperty
libcrypto.lib(libdefault-lib-winstore_store.obj) : error LNK2001: unresolved external symbol __imp_CertOpenSystemStoreW
libprotobuf-lite.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(match.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
libprotobuf.lib(descriptor.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
libprotobuf.lib(any_lite.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_search_1
GNS_Static.exe : fatal error LNK1120: 12 unresolved externals
```

However, since GameNetworkingSockets\build\vcpkg_installed\x64-windows-static\lib has only release libs, trying to use them with debug config results in mismatch between CRTs /MTd in my project and /MT in libs
UPD: my bad, debug ones are in GameNetworkingSockets\build\vcpkg_installed\x64-windows-static\debug\lib

UPD:
Adding more libs from gns cmake helps a bit, but still leaves some errors
```
target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE
        ${CMAKE_CURRENT_SOURCE_DIR}/dep/GameNetworkingSockets/lib/$<$<CONFIG:Debug>:Debug>$<$<CONFIG:Release>:Release>/GameNetworkingSockets_s.lib
        ${PB_LIB}
        ws2_32
        crypt32
)
```
```
absl_log_internal_message.lib(log_message.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
libprotobuf-lite.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_of_trivial_pos_1
libprotobuf-lite.lib(io_win32.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
absl_strings.lib(match.cc.obj) : error LNK2001: unresolved external symbol __std_find_first_of_trivial_pos_1
libprotobuf.lib(descriptor.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
libprotobuf.lib(any_lite.cc.obj) : error LNK2001: unresolved external symbol __std_find_last_trivial_1
absl_strings.lib(str_split.cc.obj) : error LNK2001: unresolved external symbol __std_search_1
GNS_Static.exe : fatal error LNK1120: 4 unresolved externals
```

**Comment by Rewriter00x** [comment]
Update: it appears those last std related errors were due to dependents from vcpkg were built with newer version than my msvc toolchain. After forcing everything to use 14.44 it now compiles with no issues. Not sure how to make it default over my 14.38 though, had to manually open vs dev console with 14.44 args
